test/e2e/restore_test.go

```diff
@@ -0,0 +1,125 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/coreos/etcd-operator/pkg/spec"
+	"github.com/coreos/etcd-operator/pkg/util/constants"
+	"github.com/coreos/etcd-operator/test/e2e/framework"
+	"github.com/coreos/etcd/clientv3"
+)
+
+func TestClusterRestore(t *testing.T) {
+	f := framework.Global
+	origEtcd := makeEtcdCluster("test-etcd-", 3)
+	testEtcd, err := createEtcdCluster(f, etcdClusterWithBackup(origEtcd, makeBackupPolicy(false)))
+	if err != nil {
+		t.Fatal(err)
+	}
+	names, err := waitUntilSizeReached(f, testEtcd.Name, 3, 60)
+	if err != nil {
+		t.Fatalf("failed to create 3 members etcd cluster: %v", err)
+	}
+
+	// set foo=bar
+	pod, err := f.KubeClient.Pods(f.Namespace.Name).Get(names[0])
+	if err != nil {
+		t.Fatal(err)
+	}
+	etcdcli, err := createEtcdClient(fmt.Sprintf("http://%s:2379", pod.Status.PodIP))
+	if err != nil {
+		t.Fatal(err)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
+	_, err = etcdcli.Put(ctx, "foo", "bar")
+	cancel()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// backup
+	if err := waitBackupPodUp(f, testEtcd.Name, 60*time.Second); err != nil {
+		t.Fatalf("failed to create backup pod: %v", err)
+	}
+	if err := makeBackup(f, testEtcd.Name); err != nil {
+		t.Fatalf("fail to make a backup: %v", err)
+	}
+	// delete cluster
+	if err := deleteEtcdCluster(f, testEtcd.Name); err != nil {
+		t.Fatal(err)
+	}
+
+	// restore
+	origEtcd.GenerateName = ""
+	origEtcd.Name = testEtcd.Name
+	origEtcd = etcdClusterWithRestore(origEtcd, &spec.RestorePolicy{
+		BackupClusterName: testEtcd.Name,
+		StorageType:       spec.BackupStorageTypePersistentVolume,
+	})
+	testEtcd, err = createEtcdCluster(f, etcdClusterWithBackup(origEtcd, makeBackupPolicy(true)))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		if err := deleteEtcdCluster(f, testEtcd.Name); err != nil {
+			t.Fatal(err)
+		}
+	}()
+	names, err = waitUntilSizeReached(f, testEtcd.Name, 3, 60)
+	if err != nil {
+		t.Fatalf("failed to create 3 members etcd cluster: %v", err)
+	}
+
+	// get foo
+	pod, err = f.KubeClient.Pods(f.Namespace.Name).Get(names[0])
+	if err != nil {
+		t.Fatal(err)
+	}
+	etcdcli, err = createEtcdClient(fmt.Sprintf("http://%s:2379", pod.Status.PodIP))
+	if err != nil {
+		t.Fatal(err)
+	}
+	ctx, cancel = context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
+	resp, err := etcdcli.Get(ctx, "foo")
+	cancel()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(resp.Kvs) != 1 {
+		t.Errorf("want 1 key result, get %d", len(resp.Kvs))
+	} else {
+		val := string(resp.Kvs[0].Value)
+		if val != "bar" {
+			t.Errorf("value want = 'foo', get = '%s'", val)
+		}
+	}
+}
+
+func createEtcdClient(addr string) (*clientv3.Client, error) {
```

Comment:
probably just MustCreateEtcdClient(t *testing.T, addr string)? Also do we have a until.go? should we move it there?